Dockerfile

```diff
@@ -1,56 +1,90 @@
-FROM php:8.0-cli as rector
-WORKDIR /rector
+ARG PHP_VERSION=8.0
+FROM php:${PHP_VERSION}-cli as base
+
+RUN apt-get update && apt-get install -y \
+    libzip4 \
+    libicu63 \
+    && rm -rf /var/lib/apt/lists/*
+
+FROM base as build
+
+WORKDIR /build
 
 # Install php extensions
 RUN apt-get update && apt-get install -y \
-        git \
-        unzip \
         g++ \
-        libzip-dev \
+        git \
         libicu-dev \
-    && rm -rf /var/lib/apt/lists/* \
-    && pecl -q install \
+        libzip-dev \
+        unzip \
+        wget \
         zip \
-    && docker-php-ext-configure \
-        opcache --enable-opcache \
-    && docker-php-ext-enable \
+    && pecl -q install \
         zip \
-        opcache \
     && docker-php-ext-configure intl \
-    && docker-php-ext-install intl
+    && docker-php-ext-configure opcache --enable-opcache \
+    && docker-php-ext-install \
+        intl \
+        opcache \
+        zip
 
 COPY --from=composer:2.0.9 /usr/bin/composer /usr/bin/composer
 
-ENV COMPOSER_ALLOW_SUPERUSER=1 COMPOSER_MEMORY_LIMIT=-1
+ENV COMPOSER_ALLOW_SUPERUSER=1 COMPOSER_MEMORY_LIMIT=-1 COMPOSER_NO_INTERACTION=1
 
-# Copy configuration
-COPY .docker/php/opcache.ini /usr/local/etc/php/conf.d/opcache.ini
+# This is to make parsing version possible
+COPY .git .git
 
+# First copy composer.json only to leverage the build cache
 COPY composer.json composer.json
-COPY stubs stubs
+RUN composer install --no-dev --no-progress --no-autoloader --prefer-dist
 
-# This is to make parsing version possible
-COPY .git .git
+# Add source and generate full autoloader
+COPY . .
+RUN composer dump-autoload --optimize --classmap-authoritative --no-dev
 
-RUN composer install --no-dev --optimize-autoloader --prefer-dist \
-    && composer clear-cache
+FROM build as build-scoped
 
-RUN mkdir /tmp/opcache
+RUN wget https://github.com/humbug/php-scoper/releases/download/0.14.0/php-scoper.phar -N --no-verbose
 
-COPY . .
+# Run php-scoper, results go to /scoped
+RUN rm -f "phpstan-for-rector.neon" \
+    && php -d memory_limit=-1 php-scoper.phar add-prefix bin config packages rules src templates vendor composer.json --output-dir /scoped --config scoper.php \
+    && composer dump-autoload --optimize --classmap-authoritative --no-dev --working-dir /scoped
 
-# To warmup opcache a little
-RUN bin/rector list
+COPY ./scoped /scoped/
 
-RUN chmod 777 -R /tmp
+FROM base as runtime
```

Comment:
Im not sure why do we need this extra `runtime` target, why it is not directly `FROM base as rector` ?

Reply:
Done in #5667